Add support for ft/syntax bash

mqudsi opened the pull request:
Currently, there is no language/filetype/syntax called "bash", it's the
sh syntax/filetype with a buffer-local variable `b:is_bash` that
determines whether or not bash-specific features are enabled when
processing the script.

The problem is that a) `set [ft|syntax]=bash` does not work, and more
importantly, b) as a result, it is not possible to set the filetype to
bash via a modeline, as `#vim: let b:is_bash=1` isn't legal.

This patch will let `set ft=bash` or `set syntax=bash` act as if
`let b:is_bash=1` were defined and then `set ft=sh` or `set syntax=sh`
was called, i.e.:

    set ft=bash
    let b:is_bash "returns 1
    echo &ft "returns sh

mqudsi committed:
Add support for ft/syntax bash

Currently, there is no language/filetype/syntax called "bash", it's the
sh syntax/filetype with a buffer-local variable `b:is_bash` that
determines whether or not bash-specific features are enabled when
processing the script.

The problem is that a) `set [ft|syntax]=bash` does not work, and more
importantly, b) as a result, it is not possible to set the filetype to
bash via a modeline, as `#vim: let b:is_bash=1` isn't legal.

This patch will let `set ft=bash` or `set syntax=bash` act as if
`let b:is_bash=1` were defined and then `set ft=sh` or `set syntax=sh`
was called, i.e.:

    set ft=bash
    let b:is_bash "returns 1
    echo &ft "returns sh

mqudsi commented:
This is marked as WIP because (despite what I mention in the changelog) `#vim: ft=bash` isn't working properly, it sets `ft=sh` correctly, but for some reason `b:is_bash` remains undefined. Is the modeline not executed as if `set ft=bash` were typed into the command buffer directly?

ref neovim/neovim#8209

chrisbra commented:
I am not sure this is so much better than what we have today. BTW: what is the reason for the empty syntax file? Are you going to maintain the filetype and syntax files then?

brammool commented:
> Currently, there is no language/filetype/syntax called "bash", it's the
> sh syntax/filetype with a buffer-local variable `b:is_bash` that
> determines whether or not bash-specific features are enabled when
> processing the script.
> 
> The problem is that a) `set [ft|syntax]=bash` does not work, and more
> importantly, b) as a result, it is not possible to set the filetype to
> bash via a modeline, as `#vim: let b:is_bash=1` isn't legal.
> 
> This patch will let `set ft=bash` or `set syntax=bash` act as if
> `let b:is_bash=1` were defined and then `set ft=sh` or `set syntax=sh`
> was called, i.e.:
> 
>     set ft=bash
>     let b:is_bash "returns 1
>     echo &ft "returns sh

I do think that supporting ":set ft=bash" is a good idea.  But then
having the value set to "sh" again is very confusing.

Can we make the plugins set the b:is_bash variable and then source the
"sh" plugin files?

-- 
Momento mori, ergo carpe diem

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

mqudsi commented:
Sorry about the empty syntax file. I'm trying to remember if that's on purpose or not (this was a month ago).

@brammool that's exactly what this does, no? I haven't actually defined a syntax for `bash`, I just set `b:is_bash` and then apply the `sh` ft and syntax. I agree that having `echo &ft` or `echo &syntax` return `bash` makes more sense then returning `sh` afterwards - any suggestions on how best to achieve that?

I don't want to maintain a separate `bash` syntax file, the `sh` syntax file with `b:is_bash` does a good job.

tonymec commented:
What about a `syntax/bash.vim` containing
``` vim
" Dummy syntax file for bash
let b:is_bash = 1
runtime! syntax/sh.vim
noautocmd setlocal syntax=bash
```
or similar (and likewise for `ftplugin/bash.vim`)? No need to change anything in the C code I think. (Look how `vim` syntax sources the scripts for the various interpreted scripts, or how `html` syntax sources the scripts for `css` and `javascript`)

mqudsi committed:
Use noautocmd to keep ft/syntax as `bash` while using `sh`

Thanks @tonymec

mqudsi committed:
Interpret bash shebang as ft=bash instead of ft=sh

Internally bash.vim will take care of calling the sh.vim runtime files
to use it for syntax and structure, but as far as the user is concerned,
this is bash.

mqudsi commented:
@tonymec I did not know about `noautocmd`, that helps tremendously!

I've updated the branch and I'm now relying on that to keep up the user-facing facade. I've additionally updated `ft.vim` to set `ft` to `bash` on a bash shebang instead of `sh`.

mqudsi commented:
_I removed the WIP tag from the title and believe this is ready for review_

brammool commented:
> _I removed the WIP tag from the title and believe this is ready for review_

I know that using filetype "sh" for ksh, bash and other types of shell
is a bit strange, and that perhaps "bash" deserves to be a separate
filetype, but I'm afraid making this change will annoy a lot of users
who rely on bash files to have filetype "sh".

In a way a bash file is both "sh" and "bash".  Unfortulately we can't
have two filetypes for one file.

At the same time, I haven't heard complaints from the current mechanism
to use filetype "sh" with b:is_bash set.  Is there really a reason to
make this change?


-- 
hundred-and-one symptoms of being an internet addict:
150. You find yourself counting emoticons to get to sleep.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

tonymec commented:
All those shell scripts have multiple common treats, and each of them is
set apart by a few quirks. As long as it is possible to handle those quirks
as various "feature sets" I'm not opposed to treating them as basically one
"language family".

As an analogy from a different language: XHTML is both XML and HTML, and
indeed it was intentionally created so that it would be parseable as either
of them. In addition, it introduces the notion of XML namespaces into HTML,
which lacked them, and this allows "mixing" HTML with theoretically "any"
XML sublanguage (e.g. IIUC MathML or, insofar as it isn't dead, XUL). I
suppose these "elements from different languages" can be handled in a
manner parallel to the handling of Perl, Python2, Python3, Scheme, Ruby and
TCL in Vim scripts, or indeed of CSS in HTML pages; but how do we handle
the fact that XHTML is "both" XML and HTML and could be "validly" parsed as
either?

Best regards,
Tony.

mqudsi commented:
Just for some background, the only reason I went down this rabbit hole at all is just to have some way to tell vim via a modeline that bash syntax is legal for the file in question.

@brammool I get what you are saying. If it’s worth anything, my goal was to come as close as possible to creating a fake ft ‘bash’ that would set the actual ft to ‘sh’ and flip the ‘bash’ flag, in an attempt to work around that particular limitation.

brammool commented:
> Just for some background, the only reason I went down this rabbit hole
> at all is just to have some way to tell vim via a modeline that bash
> syntax is legal for the file in question.
> 
> @brammool I get what you are saying. If it’s worth anything, my goal
> was to come as close as possible to creating a fake ft ‘bash’ that
> would set the actual ft to ‘sh’ and flip the ‘bash’ flag, in an
> attempt to work around that particular limitation. 

Yeah, but setting 'filetype' in the filetype plugin directly won't work
properly, because the script is sourced through an autocommand, which
doesn't work properly recursively.

Perhaps this works well enough, used as ftplugin/bash.vim:


if exists("b:did_ftplugin")
  finish
endif

let b:is_bash = 1
if exists("b:is_sh")
  unlet b:is_sh
endif
if exists("b:is_kornshell")
  unlet b:is_kornshell
endif

" Setting 'filetype' here directly won't work, since we are being invoked
" through an autocommand.  Do it later, on the BufWinEnter event.
augroup bash_filetype
  au BufWinEnter * call SetBashFt()
augroup END

func SetBashFt()
  au! bash_filetype
  set ft=sh
endfunc

-- 
"Hit any key to continue" is a lie.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

brammool commented:
Any response to my suggestion from January 12?

mqudsi commented:
Yes, two things. 

1) With the initial approach, `ft=bash` means set `ft` to `sh` with `b:is_bash=1`. Your first comment was about people expecting bash to be ft sh and still would be in that case, just ft bash is shorthand for setting both values in one go, and &ft backwards compatibility is preserved. I guess we have to decide if you want that or if you prefer that after a `let ft=bash` a subsequent `let &ft` should return `sh` (for backwards compatibility) or `bash` (for consistency).
2) Regarding “won’t work” - I’m not sure what exactly you were referring to, as the PR as it currently stands “works” but again that depends on the answer to the previous point.

brammool commented:
Well, considering how many people struggle with the current way, that is call all kinds of shell script "sh" and then using variables to say what kind of shell, I tend to lean towards having the 'filetype' be specific for the kind of shell.  And then re-using the common implementation.  This will not be backwards though, I'm not sure how we can work around that.  We could ask the user to set a preference, but then new users would still be stuck in the old way. Unless we make the preference default to the new way, then existing users may need to take action.

mqudsi commented:
I guess if `is_bash` is set before `ft=sh` is assigned, you could “upgrade” to `ft=bash` for backwards compatibility, but that may be too much magic.

brammool commented:
We actually already have ftplugin/bash.vim.  Let me add syntax/bash.vim and indent/bash.vim, so that you can do "set ft=bash".  If there are problems with that we can probably fix them.

olmokramer commented:
Vim allows for multiple filetypes separated by a `.`, so is it not possible to do something to the effect of `setfiletype sh.bash`? I use a similar technique for Vue single component files (which are basically HTML files with a few extras) in `~/.vim/filetype.vim`:

    autocmd BufRead,BufNewFile *.vue setfiletype html.vue

This loads the HTML configuration and allows me to selectively override things for Vue files without having to have a copy of the entire configuration.

brammool commented:
> Vim allows for multiple filetypes separated by a `.`, so is it not possible to do something to the effect of `setfiletype sh.bash`? I use a similar technique for Vue single component files (which are basically HTML files with a few extras) in `~/.vim/filetype.vim`:
> 
>     autocmd BufRead,BufNewFile *.vue setfiletype html.vue
> 
> This loads the HTML configuration and allows me to selectively
> override things for Vue files without having to have a copy of the
> entire configuration.

The dot notation loads two filetypes in sequence.  This doesn't work for
"sh" used for bash, because the bash-specific things are in sh.vim.
It's not something added on top of shell syntax, it changes the syntax.

-- 
ARTHUR:  Be quiet!
DENNIS:  Well you can't expect to wield supreme executive power just 'cause
         some watery tart threw a sword at you!
ARTHUR:  Shut up!
DENNIS:  I mean, if I went around sayin' I was an empereror just because some
         moistened bint had lobbed a scimitar at me they'd put me away!
                                  The Quest for the Holy Grail (Monty Python)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

brammool commented:
Since this is incomplete and there is no progress, let me close this. I would appreciated a solution that works for "everybody".
